File: src/frontend/parser/expressionParser.ts
// imports 
import { TokenType } from "../lexer/types";
import { ExpressionKind, Expression, NumericLiteral, Identifier } from "./types";
import ParserContext from "./parserContext";

const ASSIGNMENT_OPERATORS = new Set(["=", "+=", "-=", "*=", "/="]);
const EQUALITY_OPERATORS = new Set(["===", "!=="]);
const COMPARISON_OPERATORS = new Set([">", "<", ">=", "<="]);
const ADDITIVE_OPERATORS = new Set(["+", "-"]);
const MULTIPLICATIVE_OPERATORS = new Set(["*", "/", "%"]);

export default class ExpressionParser {
  private parserContext: ParserContext;

  constructor(ctx: ParserContext){
    this.parserContext = ctx;
  }

  public parseExpression(): Expression {
    return this.parseAssignmentExpression();
  }

  public parseAssignmentExpression(): Expression {
    const left = this.parseLogicalOrExpression();

    if(ASSIGNMENT_OPERATORS.has(this.parserContext.peek().value)){
      // TODO: Later left side can be an assignable expression like member expression obj.x (identifier will work now)

      if(left.kind !== ExpressionKind.Identifier){
        throw new Error("Invalid syntax, cannot assign a expression to anything other than identifier");
      }

      const op = this.parserContext.eat().value; // eat assignemnt token 
      return { kind: ExpressionKind.AssignmentExpression, left, right: this.parseAssignmentExpression(), op };
    }

    return left;
  }

  public parseLogicalOrExpression(): Expression {
    let left = this.parseLogicalAndExpression();

    
    while(this.parserContext.peek().value === "||"){
      const op = this.parserContext.eat().value;
      const right = this.parseLogicalAndExpression();

      left = {
        kind: ExpressionKind.BinaryExpression,
        left,
        right,
        op,
      }
    }

    return left;
  }

  public parseLogicalAndExpression(): Expression {
    let left = this.parseEqualityExpression();

    
    while(this.parserContext.peek().value === "&&"){
      const op = this.parserContext.eat().value;
      const right = this.parseEqualityExpression();

      left = {
        kind: ExpressionKind.BinaryExpression,
        left,
        right,
        op,
      }
    }

    return left;
  }

  public parseEqualityExpression(): Expression {
    let left = this.parseComparisonExpression();

    
    while(EQUALITY_OPERATORS.has(this.parserContext.peek().value)){
      const op = this.parserContext.eat().value;
      const right = this.parseComparisonExpression();

      left = {
        kind: ExpressionKind.BinaryExpression,
        left,
        right,
        op,
      }
    }

    return left;
  }

  public parseComparisonExpression(): Expression {
    let left = this.parseAdditiveExpression();

    
    while(COMPARISON_OPERATORS.has(this.parserContext.peek().value)){
      const op = this.parserContext.eat().value;
      const right = this.parseAdditiveExpression();

      left = {
        kind: ExpressionKind.BinaryExpression,
        left,
        right,
        op,
      }
    }

    return left;
  }

  public parseAdditiveExpression(): Expression {
    let left = this.parseMultiplicativeExpression();

    while(ADDITIVE_OPERATORS.has(this.parserContext.peek().value)){
      const op = this.parserContext.eat().value;
      const right = this.parseMultiplicativeExpression();

      left = {
        kind: ExpressionKind.BinaryExpression,
        left,
        right, 
        op,
      }
    }

    return left;
  }


  public parseMultiplicativeExpression(): Expression {
    let left = this.parseCallExpression();

    while(MULTIPLICATIVE_OPERATORS.has(this.parserContext.peek().value)){
      const op = this.parserContext.eat().value;
      const right = this.parseCallExpression();
      left = {
        kind: ExpressionKind.BinaryExpression,
        left,
        right, 
        op,
      }

    }

    return left; 
  }

  public parseCallExpression(): Expression {
    let expr = this.parsePrimaryExpression();

    while(this.parserContext.peek().type === TokenType.OPEN_PARENTHESIS){
      this.parserContext.eat(); // eat '(' 

      const args = this.parseArguments();
        expr = {
        kind: ExpressionKind.CallExpression,
        callee: expr,
        args
      }

      if(this.parserContext.peek().type === TokenType.CLOSED_PARENTHESIS){
        this.parserContext.eat();
      }
      else {
        throw new Error("No matching closed parenthesis");
      }
    }

    return expr;
  }

  public parsePrimaryExpression(): Expression {
    
    switch (this.parserContext.peek().type) {
      case TokenType.NUMBER:{
        return this.parseNumericLiteral();
      }
        
      case TokenType.IDENTIFIER: {
        return this.parseIdentifier();  
      }
    
      case TokenType.OPEN_PARENTHESIS: {
        return this.parseGroupedExpression();
      }

      default: {
        console.error(`Error!, Could not identify symbol '${this.parserContext.eat().value}'`);
        process.exit(1);
      }
    }
  }

  public parseIdentifier(): Identifier {
    if (this.parserContext.peek().type !== TokenType.IDENTIFIER) {
      throw new Error("Expected identifier");
    }

    return {
      kind: ExpressionKind.Identifier,
      symbol: this.parserContext.eat().value,
    };
  }

  public parseNumericLiteral(): NumericLiteral {
    if(this.parserContext.peek().type !== TokenType.NUMBER){
      throw new Error("Expected number");
    }

    return {
      kind: ExpressionKind.NumericLiteral,
      value: parseFloat(this.parserContext.eat().value),
    };
  }

  public parseGroupedExpression(): Expression {
    if(this.parserContext.peek().type !== TokenType.OPEN_PARENTHESIS){
      throw new Error("Expected open parenthesis");
    }

    this.parserContext.eat(); // '('
    const expr = this.parseExpression();

    if (this.parserContext.peek().type !== TokenType.CLOSED_PARENTHESIS) {
      throw new Error("Non-matching parenthesis");
    }

    this.parserContext.eat(); // ')'
    return expr;
  }

  public parseArguments(): Expression[] {
    const args: Expression[] = [];

    if(this.parserContext.peek().type === TokenType.CLOSED_PARENTHESIS){
      return args;
    }

    args.push(this.parseExpression());

    while(!this.parserContext.endOf() && this.parserContext.peek().type === TokenType.COMMA){
      this.parserContext.eat(); // eat comma 
      args.push(this.parseExpression());
    }
    
    if(this.parserContext.peek().type !== TokenType.CLOSED_PARENTHESIS){
      throw new Error("No matching closed parenthesis")
    }
    
    return args;
  } 
}

// priority order of operators in expression parsing  
// 1. Statement Execution
// 2. Assignment Operator 
// 3. Logical Operators
// 4. Comparison Operators 
// 5. Additive Operators 
// 6. Multiplicative Operators 
// 7. Call Expression 
// 8. Primary expression

File: src/frontend/parser/expression_parser.cpp
#include <stdexcept>
#include <unordered_set>

#include <scriptum/frontend/parser/expression_parser.hpp>
#include <scriptum/frontend/lexer/token.hpp>

namespace scriptum {

namespace {

const std::unordered_set<std::string> kAssignmentOperators = {"=", "+=", "-=", "*=", "/="};
const std::unordered_set<std::string> kEqualityOperators = {"===", "!=="};
const std::unordered_set<std::string> kComparisonOperators = {">", "<", ">=", "<="};
const std::unordered_set<std::string> kAdditiveOperators = {"+", "-"};
const std::unordered_set<std::string> kMultiplicativeOperators = {"*", "/", "%"};

} // namespace

ExpressionParser::ExpressionParser(ParserContext& ctx) : m_ctx(ctx) {}

ExprPtr ExpressionParser::parseExpression()
{
    return parseAssignmentExpression();
}

ExprPtr ExpressionParser::parseAssignmentExpression()
{
    auto left = parseLogicalOrExpression();

    if (kAssignmentOperators.contains(m_ctx.peek().value))
    {
        const Identifier* const id = std::get_if<Identifier>(&left->node);
        if (id == nullptr)
        {
            throw std::runtime_error(
                "Invalid syntax, cannot assign a expression to anything other than identifier");
        }

        const std::string op = m_ctx.eat().value;
        return std::make_unique<Expression>(std::in_place_type<AssignmentExpression>, *id,
                                            parseAssignmentExpression(), op);
    }

    return left;
}

ExprPtr ExpressionParser::parseLogicalOrExpression()
{
    auto left = parseLogicalAndExpression();

    while (m_ctx.peek().value == "||")
    {
        const std::string op = m_ctx.eat().value;
        auto right = parseLogicalAndExpression();
        left = std::make_unique<Expression>(std::in_place_type<BinaryExpression>, std::move(left),
                                            std::move(right), op);
    }

    return left;
}

ExprPtr ExpressionParser::parseLogicalAndExpression()
{
    auto left = parseEqualityExpression();

    while (m_ctx.peek().value == "&&")
    {
        const std::string op = m_ctx.eat().value;
        auto right = parseEqualityExpression();
        left = std::make_unique<Expression>(std::in_place_type<BinaryExpression>, std::move(left),
                                            std::move(right), op);
    }

    return left;
}

ExprPtr ExpressionParser::parseEqualityExpression()
{
    auto left = parseComparisonExpression();

    while (kEqualityOperators.contains(m_ctx.peek().value))
    {
        const std::string op = m_ctx.eat().value;
        auto right = parseComparisonExpression();
        left = std::make_unique<Expression>(std::in_place_type<BinaryExpression>, std::move(left),
                                            std::move(right), op);
    }

    return left;
}

ExprPtr ExpressionParser::parseComparisonExpression()
{
    auto left = parseAdditiveExpression();

    while (kComparisonOperators.contains(m_ctx.peek().value))
    {
        const std::string op = m_ctx.eat().value;
        auto right = parseAdditiveExpression();
        left = std::make_unique<Expression>(std::in_place_type<BinaryExpression>, std::move(left),
                                            std::move(right), op);
    }

    return left;
}

ExprPtr ExpressionParser::parseAdditiveExpression()
{
    auto left = parseMultiplicativeExpression();

    while (kAdditiveOperators.contains(m_ctx.peek().value))
    {
        const std::string op = m_ctx.eat().value;
        auto right = parseMultiplicativeExpression();
        left = std::make_unique<Expression>(std::in_place_type<BinaryExpression>, std::move(left),
                                            std::move(right), op);
    }

    return left;
}

ExprPtr ExpressionParser::parseMultiplicativeExpression()
{
    auto left = parseCallExpression();

    while (kMultiplicativeOperators.contains(m_ctx.peek().value))
    {
        const std::string op = m_ctx.eat().value;
        auto right = parseCallExpression();
        left = std::make_unique<Expression>(std::in_place_type<BinaryExpression>, std::move(left),
                                            std::move(right), op);
    }

    return left;
}

ExprPtr ExpressionParser::parseCallExpression()
{
    auto expr = parsePrimaryExpression();

    while (m_ctx.peek().type == TokenType::OpenParenthesis)
    {
        m_ctx.eat();
        auto args = parseArguments();
        expr = std::make_unique<Expression>(std::in_place_type<CallExpression>, std::move(expr),
                                            std::move(args));

        if (m_ctx.peek().type == TokenType::ClosedParenthesis)
        {
            m_ctx.eat();
        }
        else
        {
            throw std::runtime_error("No matching closed parenthesis");
        }
    }

    return expr;
}

ExprPtr ExpressionParser::parsePrimaryExpression()
{
    switch (m_ctx.peek().type)
    {
        case TokenType::Number:
            return parseNumericLiteral();
        case TokenType::Identifier: {
            Identifier id = parseIdentifier();
            return std::make_unique<Expression>(std::in_place_type<Identifier>,
                                                std::move(id.symbol));
        }
        case TokenType::OpenParenthesis:
            return parseGroupedExpression();
        case TokenType::Eof:
            throw std::runtime_error("Erorr!, Invalid primary expression in statement");
        default:
            throw std::runtime_error("Error!, Could not identify symbol '" + m_ctx.eat().value +
                                     "'");
    }
}

Identifier ExpressionParser::parseIdentifier()
{
    if (m_ctx.peek().type != TokenType::Identifier)
    {
        throw std::runtime_error("Expected identifier");
    }
    return Identifier(m_ctx.eat().value);
}

ExprPtr ExpressionParser::parseNumericLiteral()
{
    if (m_ctx.peek().type != TokenType::Number)
    {
        throw std::runtime_error("Expected number");
    }
    return std::make_unique<Expression>(std::in_place_type<NumericLiteral>,
                                        std::stod(m_ctx.eat().value));
}

ExprPtr ExpressionParser::parseGroupedExpression()
{
    if (m_ctx.peek().type != TokenType::OpenParenthesis)
    {
        throw std::runtime_error("Expected open parenthesis");
    }

    m_ctx.eat();
    auto expr = parseExpression();

    if (m_ctx.peek().type != TokenType::ClosedParenthesis)
    {
        throw std::runtime_error("Non-matching parenthesis");
    }

    m_ctx.eat();
    return expr;
}

std::vector<ExprPtr> ExpressionParser::parseArguments()
{
    std::vector<ExprPtr> args;

    if (m_ctx.peek().type == TokenType::ClosedParenthesis)
    {
        return args;
    }

    args.push_back(parseExpression());

    while (!m_ctx.endOf() && m_ctx.peek().type == TokenType::Comma)
    {
        m_ctx.eat();
        args.push_back(parseExpression());
    }

    if (m_ctx.peek().type != TokenType::ClosedParenthesis)
    {
        throw std::runtime_error("No matching closed parenthesis");
    }

    return args;
}

} // namespace scriptum
